pkg/services/meta/container.go

```diff
@@ -0,0 +1,129 @@
+	objectsdk "github.com/nspcc-dev/neofs-sdk-go/object"
+)
+
+type containerStorage struct {
+	m        sync.RWMutex
+	opsBatch map[string][]byte
+
+	path string
+	mpt  *mpt.Trie
+	db   storage.Store
+}
+
+func (s *containerStorage) drop() error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	err := s.db.Close()
+	if err != nil {
+		return fmt.Errorf("close container storage: %w", err)
+	}
+
+	err = os.RemoveAll(s.path)
+	if err != nil {
+		return fmt.Errorf("remove container storage: %w", err)
+	}
+
+	return nil
+}
+
+func (s *containerStorage) putObject(e objEvent) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	newKVs := make(map[string][]byte)
+	commsuffix := e.oID[:]
+
+	// batching that is implemented for MPT ignores key's first byte
+
+	newKVs[string(append([]byte{0, oidIndex}, commsuffix...))] = []byte{}
+	newKVs[string(append([]byte{0, sizeIndex}, commsuffix...))] = e.size.Bytes()
+	if len(e.firstObject) == sha256.Size {
+		newKVs[string(append([]byte{0, firstPartIndex}, commsuffix...))] = e.firstObject
+	}
+	if len(e.prevObject) == sha256.Size {
+		newKVs[string(append([]byte{0, previousPartIndex}, commsuffix...))] = e.prevObject
+	}
+	if len(e.deletedObjects) > 0 {
+		newKVs[string(append([]byte{0, deletedIndex}, commsuffix...))] = e.deletedObjects
+	}
+	if len(e.lockedObjects) > 0 {
+		newKVs[string(append([]byte{0, lockedIndex}, commsuffix...))] = e.lockedObjects
+	}
+	if e.typ != objectsdk.TypeRegular {
+		newKVs[string(append([]byte{0, typeIndex}, commsuffix...))] = []byte{byte(e.typ)}
+	}
+
+	if s.opsBatch == nil {
+		s.opsBatch = make(map[string][]byte)
+	}
+	maps.Copy(s.opsBatch, newKVs)
+
+	err := s.db.PutChangeSet(mptToStoreBatch(newKVs), nil)
+	if err != nil {
+		return fmt.Errorf("put MPT KVs to the raw storage manually: %w", err)
+	}
+
+	// TODO: receive full object header and store its non-MPT parts in
+	//  persistent object storage
+
+	return nil
+}
+
+// mptToStoreBatch drops the first byte from every key in the map.
+func mptToStoreBatch(b map[string][]byte) map[string][]byte {
+	for k, v := range b {
+		b[k[1:]] = v
+	}
+
+	return b
+}
+
+func storageForContainer(rootPath string, cID cid.ID) (*containerStorage, error) {
+	cIDString := cID.EncodeToString()
+	p := path.Join(rootPath, cIDString)
+
+	st, err := storage.NewBoltDBStore(dbconfig.BoltDBOptions{FilePath: p, ReadOnly: false})
+	if err != nil {
+		return nil, fmt.Errorf("open bolt store at %q: %w", p, err)
+	}
+
+	var prevRootNode mpt.Node
+	root, err := st.Get([]byte{rootKey})
+	if !errors.Is(err, storage.ErrKeyNotFound) {
+		if err != nil {
+			return nil, fmt.Errorf("get state root from db: %w", err)
+		}
+
+		if len(root) != util.Uint256Size {
+			return nil, fmt.Errorf("root hash from db is %d bytes long, expect %d", len(root), util.Uint256Size)
+		}
+
+		prevRootNode = mpt.NewHashNode([util.Uint256Size]byte(root))
+	}
+
+	mTrie := mpt.NewTrie(prevRootNode, mpt.ModeLatest, storage.NewMemCachedStore(st))
+
+	return &containerStorage{
+		m:    sync.RWMutex{},
```

Comment:
no need

Reply:
i prefer explicit struct initiation but searched for this in our code and there is not a single second example like this so ok, dropped